### DISCUSSION TITLE
ESP32 Bluetooth transmitting power

### DISCUSSION THREAD

**Discussion by louis12A** [micropython/micropython#10615] [General]
Hi

I want to know if it is possible to increase the dBm output of the esp32 wroom using Arduino ide.
From research it shows it's only possible with ESP IDF, if so then what is the default output dBm when you buy the module.
I can't seem to find enough info on this matter.

**Comment by jimmo** [comment]
We do not currently provide an API to set this.

> I can't seem to find enough info on this matter.

https://espressif-docs.readthedocs-hosted.com/projects/espressif-esp-faq/en/latest/software-framework/ble-bt.html#what-is-the-default-bluetooth-transmit-power-for-esp32-sdk

**Comment by pidou46** [comment]
Hi,
Is there a way to set it by another way ? at runtime or at least configuration file at firmware building time ?
I think it would be a nice addition to the mp API thought (to save battery or to build long range beacon)

**Reply by cve2022** [reply]
> Hi, Is there a way to set it by another way ? at runtime or at least configuration file at firmware building time ? I think it would be a nice addition to the mp API thought (to save battery or to build long range beacon)

If you know what registers are accessed to change the transmitting power, you may try to access them directly using [mem32](https://docs.micropython.org/en/latest/library/machine.html). There are mem32 usage examples [for the ESP32](https://docs.micropython.org/en/latest/esp32/tutorial/peripheral_access.html).

**Reply by jimmo** [reply]
You can call the same function you'd use from Arduino. https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/bluetooth/controller_vhci.html#_CPPv420esp_ble_tx_power_set20esp_ble_power_type_t17esp_power_level_t

Either from ports/esp32/mpnimbleport.c or from extmod/nimble/modbluetooth_nimble.c

**Reply by pidou46** [reply]
Thanks for your suggestions

The solution suggested by cve2022 would be nice for me because it doesn't imply firmware compilation, but I haven't found registry address in the esp32_technical_reference_manual_en.pdf

**Reply by cve2022** [reply]
> The solution suggested by cve2022 would be nice for me because it doesn't imply firmware compilation, but I haven't found registry address in the esp32_technical_reference_manual_en.pdf

I took a quick look at the IDF source code out of curiosity. I thought one could look at the C source code and reproduce it in MP but it looks like the power setting is in a precompiled library, the C code just calls a function in the libray. If that is really the case, then my suggestion, unfortunately, would not work, considering the documentation does not have info on the bluetooth module. 

On second thought, is very likely the wireless module is external to the processor and communication with it is through SPI so there is not direct register control of BT/WiFi in the same sense as with other peripherals.

**Comment by Lobo-T** [comment]
Here is a thread about this on the old forum: https://forum.micropython.org/viewtopic.php?p=62588#p62588
